Revert not exporting rpaths to CMakeCache in CachedCMakePackages

adrienbernede opened the pull request:
When using the CMakeCache file generated by a CachedCMakePackage without the Spack compiler wrappers, some rpaths are missing:
- the extra-rpaths set in the compilers.yaml
- the implicit link directories from the compiler

This resulted in missing rpaths when one or the other was needed.

@haampie this was tested and effectively fixes my builds.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

adrienbernede committed:
Revert not exporting rpaths to CMakeCache in CachedCMakePackages

When using the CMakeCache file generated by a CachedCMakePackage
without the Spack compiler wrappers, some rpaths are missing:
- the extra-rpaths set in the compilers.yaml
- the implicit link directories from the compiler

This resulted in missing rpaths when one or the other was needed.

haampie commented on lib/spack/spack/build_environment.py:
```diff
@@ -789,6 +789,21 @@ def get_rpath_deps(pkg: spack.package_base.PackageBase) -> List[spack.spec.Spec]
     return _get_rpath_deps_from_spec(pkg.spec, pkg.transitive_rpaths)
 
 
+def get_rpaths(pkg):
```
Transitive deps are covered by `CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON`, so I would not revive this. It's missing various bits done elsewhere in `build_environment` so like this it has bugs.

adrienbernede replied:
Unfortunately, applying the suggested change is causing a failure in one of my scenario. I'll investigate tomorrow because it's not easy to state whether it's on my side or not since it involves hip support for which we have custom logic.

adrienbernede replied:
OK I'm fine with things as they are now!

adrienbernede committed:
Review fix: do not explicit rpaths to transitive deps

Transitive dependencies are handled by CMake, instructed
to do so with CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON

adrienbernede commented:
@haampie if you have the bandwith to review it today, that's awesome cause then I can get it in the next snapshot. Otherwise, so be it. Thank you.

adrienbernede committed:
Add default rpaths

adrienbernede committed:
Merge tag 'develop-2025-01-05' into woptim/revert-cached-cmake-rpath

Snapshot release 2025-01-05

adrienbernede commented:
Updated with latest snapshot release.

haampie commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -298,6 +298,17 @@ def initconfig_hardware_entries(self):
+        complete_rpath_list = ";".join(default_rpaths)
+        if "SPACK_COMPILER_EXTRA_RPATHS" in os.environ:
+            spack_extra_rpaths_env = os.environ["SPACK_COMPILER_EXTRA_RPATHS"]
+            spack_extra_rpaths_path = spack_extra_rpaths_env.replace(os.pathsep, ";")
```
```suggestion
            spack_extra_rpaths_path = spack_extra_rpaths_env.replace(":", ";")
```

haampie commented on lib/spack/spack/build_systems/cached_cmake.py:
```diff
@@ -298,6 +298,17 @@ def initconfig_hardware_entries(self):
+
+        if "SPACK_COMPILER_IMPLICIT_RPATHS" in os.environ:
+            spack_implicit_rpaths_env = os.environ["SPACK_COMPILER_IMPLICIT_RPATHS"]
+            spack_implicit_rpaths_path = spack_implicit_rpaths_env.replace(os.pathsep, ";")
```
```suggestion
            spack_implicit_rpaths_path = spack_implicit_rpaths_env.replace(":", ";")
```

adrienbernede committed:
Apply concise notation suggested by @haampie
